libraries/AC_SafeRTL/AC_SafeRTL.cpp

```diff
@@ -0,0 +1,551 @@
+        }
+        return;
+    }
+
+    if (position_ok) { // meters from origin, NED
+        // append the new point, if appropriate
+        if (HYPOT(current_pos, path[_last_index]) > SAFERTL_POSITION_DELTA) {
+            // add the breadcrumb
+            path[++_last_index] = current_pos;
+
+            if (_logging_enabled) {
+                DataFlash_Class::instance()->Log_Write_SRTL(DataFlash_Class::SRTL_POINT_ADD, current_pos);
+            }
+
+            // if cleanup algorithms are finished (And therefore not runnning), reset them
+            if (_simplification_complete) {
+                _reset_simplification(false);
+            }
+            if (_pruning_complete) {
+                _reset_pruning(false);
+            }
+        }
+    }
+}
+
+/**
+*  Run this method only when preparing to initiate the RTL procedure. Returns a
+*  pointer to the cleaned-up path. Returns nullptr if the cleanup algorithms aren't ready yet.
+*  If this happens, just run this method again a bit later.
+*
+*  Probably best not to run this unless cleanup_ready() is returning true
+*/
+Vector3f* SafeRTL_Path::thorough_cleanup()
+{
+    // apply simplification
+    _zero_points_by_simplification_bitmask();
+
+    // apply pruning
+    _zero_points_by_loops(SAFERTL_MAX_ALLOWABLE_PATH_LEN); // prune every single loop
+
+    _remove_empty_points();
+
+    // end by resetting the state of the cleanup methods.
+    _reset_simplification(true);
+    _reset_pruning(true);
+
+    return path;
+}
+
+/**
+*   Returns true if the list is empty after popping this point.
+*/
+bool SafeRTL_Path::pop_point(Vector3f& point)
+{
+    if (_last_index == 0) {
+        point = path[0];
+        return true;
+    }
+    point =  path[_last_index--];
+    return false;
+}
+
+/**
+*   Pass nullptr for the gcs argument if you don't want errors being reported to the user.
+*/
+void SafeRTL_Path::reset_path(bool position_ok, const Vector3f start, GCS* gcs)
+{
+    // if the user wishes to deactivate safertl, deactivate it.
+    if (desired_path_len == 0 || is_zero(accuracy)) {
+        _active = false;
+        return;
+    }
+
+    // reset simplification
+    _reset_simplification(true);
+    // reset pruning
+    _reset_pruning(true);
+
+    if (!position_ok) {
+        _active = false;
+        if (_logging_enabled) {
+            DataFlash_Class::instance()->Log_Write_SRTL(DataFlash_Class::SRTL_DEACTIVATED_BAD_POSITION, {0.0f, 0.0f, 0.0f});
+        }
+        if (gcs) {
+            gcs->send_text(MAV_SEVERITY_WARNING, "SafeRTL Unavailable: Bad Position");
+        }
+        return;
+    }
+
+    // constrain the path length, in case the user decided to make the path unreasonably long.
+    desired_path_len = MIN(SAFERTL_MAX_ALLOWABLE_PATH_LEN, desired_path_len);
+
+    // if it's time to change the size of the arrays, delete them, then reinitialize them
+    if (desired_path_len != _current_path_len) {
+        if (path != nullptr) {
```

Comment:
That's interesting, because SITL was actually crashing before I added those checks. The error I was getting was `free(): invalid next size (fast)`. As soon as I added the nullptr check, I didn't see that issue any more.